From: David Hildenbrand <dahi@linux.vnet.ibm.com>
Subject: virtio_blk: verify if queue is broken after virtqueue_get_buf()
Patch-mainline: v3.13-rc1
Git-commit: 7f03b17d5c3f3b63dcacf0e481cf94c9d07267ab
References: bnc#931860, LTC#VS1256

Summary:     s390/virtio: Performance enhanced paravirtualized I/O drivers
Description: This feature enhances performance of the paravirtualized I/O
             drivers on s390 with focus on virtio and virtio-ccw.

Upstream-Description:

             virtio_blk: verify if queue is broken after virtqueue_get_buf()

             In case virtqueue_get_buf() returned with a NULL pointer verify if the
             virtqueue is broken in order to leave while loop.

             Signed-off-by: Heinz Graalfs <graalfs@linux.vnet.ibm.com>
             Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>


Signed-off-by: David Hildenbrand <dahi@linux.vnet.ibm.com>
Acked-by: John Jolly <jjolly@suse.de>
---
 drivers/block/virtio_blk.c |    2 ++
 1 file changed, 2 insertions(+)

--- a/drivers/block/virtio_blk.c
+++ b/drivers/block/virtio_blk.c
@@ -292,6 +292,8 @@ static void virtblk_done(struct virtqueu
 				req_done = true;
 			}
 		}
+		if (unlikely(virtqueue_is_broken(vq)))
+			break;
 	} while (!virtqueue_enable_cb(vq));
 	/* In case queue is stopped waiting for more buffers. */
 	if (req_done)
